libp2p/utils/zeroqueue.nim

```diff
@@ -0,0 +1,76 @@
+# Nim-Libp2p
+# Copyright (c) 2025 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+type Chunk = object
+  address: ptr byte
+  size: int
+  start: int
+
+proc len(c: Chunk): int {.inline.} =
+  return c.size - c.start
+
+type ZeroQueue* = object
+  # ZeroQueue is queue structure optimized for efficient pushing and popping of 
+  # byte sequences `seq[byte]` (called chunks). This type is useful for streaming or buffering 
+  # scenarios where chunks of binary data are accumulated and consumed incrementally.
+  chunks: seq[Chunk]
+
+proc clear*(q: var ZeroQueue) =
+  q.chunks = @[]
+
+proc isEmpty*(q: ZeroQueue): bool =
+  return q.chunks.len == 0
+
+proc len*(q: ZeroQueue): int64 =
+  var l: int64
+  for b in q.chunks:
+    l += b.len()
+  return l
+
+proc push*(q: var ZeroQueue, b: seq[byte]) =
+  if b.len > 0:
+    q.chunks.add(Chunk(address: addr b[0], size: b.len, start: 0))
```

Comment:
this idea unfortunately will lead to a crash or invalid data. if `b` is garbage-collected before the chunk is removed, the program will crash (or write some random data). 

Generally, `ptr` is used when the lifetime of the pointee is managed by "something", such as a `ref`, and that something outlives the `ptr` itself.